File: lib/src/facts/windows/networking_resolver.cc
#include <internal/facts/windows/networking_resolver.hpp>
#include <leatherman/windows/registry.hpp>
#include <leatherman/windows/system_error.hpp>
#include <internal/util/windows/wsa.hpp>
#include <leatherman/windows/windows.hpp>
#include <leatherman/logging/logging.hpp>
#include <boost/range/combine.hpp>
#include <boost/nowide/convert.hpp>
#include <iomanip>
#include <Ws2tcpip.h>
#include <iphlpapi.h>

#ifdef interface
  // Something's bleeding in and making it impossible to instantiate an interface object.
  #undef interface
#endif

using namespace std;
using namespace facter::util;
using namespace facter::util::windows;
using namespace leatherman::windows;

namespace facter { namespace facts { namespace windows {

    networking_resolver::networking_resolver()
    {
    }

    static string get_computername(COMPUTER_NAME_FORMAT nameFormat)
    {
        DWORD size = 0u;
        GetComputerNameExW(nameFormat, nullptr, &size);
        if (GetLastError() != ERROR_MORE_DATA) {
            LOG_DEBUG("failure resolving hostname: {1}", leatherman::windows::system_error());
            return "";
        }

        wstring buffer(size, '\0');
        if (!GetComputerNameExW(nameFormat, &buffer[0], &size)) {
            LOG_DEBUG("failure resolving hostname: {1}", leatherman::windows::system_error());
            return "";
        }

        buffer.resize(size);
        return boost::nowide::narrow(buffer);
    }

    sockaddr_in networking_resolver::create_ipv4_mask(uint8_t masklen)
    {
        sockaddr_in mask = {AF_INET};
        if (ConvertLengthToIpv4Mask(masklen, &mask.sin_addr.S_un.S_addr) != NO_ERROR) {
            LOG_DEBUG("failed creating IPv4 mask of length {1}", masklen);
        }
        return mask;
    }

    sockaddr_in6 networking_resolver::create_ipv6_mask(uint8_t masklen)
    {
        sockaddr_in6 mask = {AF_INET6};
        const uint8_t incr = 32u;
        for (size_t i = 0; i < 16 && masklen > 0; i += 4, masklen -= min(masklen, incr)) {
            if (ConvertLengthToIpv4Mask(min(masklen, incr), reinterpret_cast<PULONG>(&mask.sin6_addr.u.Byte[i])) != NO_ERROR) {
                LOG_DEBUG("failed creating IPv6 mask with component of length {1}", incr);
                break;
            }
        }
        return mask;
    }

    sockaddr_in networking_resolver::mask_ipv4_address(sockaddr const* addr, sockaddr_in const& mask)
    {
        sockaddr_in masked = *reinterpret_cast<sockaddr_in const*>(addr);
        masked.sin_addr.S_un.S_addr &= mask.sin_addr.S_un.S_addr;
        return masked;
    }

    sockaddr_in6 networking_resolver::mask_ipv6_address(sockaddr const* addr, sockaddr_in6 const& mask)
    {
        sockaddr_in6 masked = *reinterpret_cast<sockaddr_in6 const*>(addr);
        for (size_t i = 0; i < 8; ++i) {
            masked.sin6_addr.u.Word[i] &= mask.sin6_addr.u.Word[i];
        }
        return masked;
    }

    networking_resolver::data networking_resolver::collect_data(collection& facts)
    {
        data result;

        result.hostname = get_computername(ComputerNameDnsHostname);

        try {
            result.domain = registry::get_registry_string(registry::HKEY::LOCAL_MACHINE,
                "SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters\\", "Domain");
        } catch (registry_exception &e) {
            LOG_DEBUG("failure getting networking::domain fact: {1}", e.what());
        }

        // Get linked list of adapters.
        ULONG family = AF_UNSPEC;
        ULONG flags = GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_DNS_SERVER;

        // Pre-allocate and try several times, because the adapter configuration may change between calls.
        ULONG outBufLen = 15000;
        vector<char> pAddresses(outBufLen);
        DWORD err;
        for (int i = 0; i < 3; ++i) {
            err = GetAdaptersAddresses(family, flags, nullptr,
                reinterpret_cast<PIP_ADAPTER_ADDRESSES>(pAddresses.data()), &outBufLen);
            if (err == ERROR_SUCCESS) {
                break;
            } else if (err == ERROR_BUFFER_OVERFLOW) {
                pAddresses.resize(outBufLen);
            } else {
                LOG_DEBUG("failure resolving networking facts: {1}", leatherman::windows::system_error(err));
                return result;
            }
        }

        if (err != ERROR_SUCCESS) {
            LOG_DEBUG("failure resolving networking facts: {1}", leatherman::windows::system_error(err));
            return result;
        }

        facter::util::windows::wsa winsock;

        for (auto pCurAddr = reinterpret_cast<PIP_ADAPTER_ADDRESSES>(pAddresses.data());
            pCurAddr; pCurAddr = pCurAddr->Next) {
            if (pCurAddr->OperStatus != IfOperStatusUp ||
                (pCurAddr->IfType != IF_TYPE_ETHERNET_CSMACD && pCurAddr->IfType != IF_TYPE_IEEE80211)) {
                continue;
            }

            if (result.domain.empty()) {
                // If domain isn't set in the registry, fall back to the first DnsDomain encountered in the adapters.
                result.domain = boost::nowide::narrow(pCurAddr->DnsSuffix);
            }

            interface net_interface;
            net_interface.name = boost::nowide::narrow(pCurAddr->FriendlyName);

            // Only supported on platforms after Windows Server 2003.
            if (pCurAddr->Flags & IP_ADAPTER_DHCP_ENABLED && pCurAddr->Length >= sizeof(IP_ADAPTER_ADDRESSES_LH)) {
                auto adapter = reinterpret_cast<IP_ADAPTER_ADDRESSES_LH&>(*pCurAddr);
                if (adapter.Flags & IP_ADAPTER_DHCP_ENABLED) {
                    try {
                        net_interface.dhcp_server = winsock.saddress_to_string(adapter.Dhcpv4Server);
                    } catch (wsa_exception &e) {
                        LOG_DEBUG("failed to retrieve dhcp v4 server address for {1}: {2}", net_interface.name, e.what());
                    }
                }
            }

            for (auto it = pCurAddr->FirstUnicastAddress; it; it = it->Next) {
                string addr;
                try {
                    addr = winsock.saddress_to_string(it->Address);
                } catch (wsa_exception &e) {
                    string iptype =
                        (it->Address.lpSockaddr->sa_family == AF_INET) ? " v4"
                        : (it->Address.lpSockaddr->sa_family == AF_INET6) ? " v6"
                        : "";
                    LOG_DEBUG("failed to retrieve ip{1} address for {2}: {3}",
                        iptype, net_interface.name, e.what());
                }

                if (addr.empty()) {
                    continue;
                }

                if (it->Address.lpSockaddr->sa_family == AF_INET || it->Address.lpSockaddr->sa_family == AF_INET6) {
                    bool ipv6 = it->Address.lpSockaddr->sa_family == AF_INET6;

                    binding b;
                    b.address = addr;

                    // Need to do lookup based on the structure length.
                    auto adapterAddr = reinterpret_cast<IP_ADAPTER_UNICAST_ADDRESS_LH&>(*it);
                    if (ipv6) {
                        auto mask = create_ipv6_mask(adapterAddr.OnLinkPrefixLength);
                        auto masked = mask_ipv6_address(it->Address.lpSockaddr, mask);
                        b.netmask = winsock.address_to_string(mask);
                        b.network = winsock.address_to_string(masked);
                    } else {
                        auto mask = create_ipv4_mask(adapterAddr.OnLinkPrefixLength);
                        auto masked = mask_ipv4_address(it->Address.lpSockaddr, mask);
                        b.netmask = winsock.address_to_string(mask);
                        b.network = winsock.address_to_string(masked);
                    }

                    if (ipv6) {
                        net_interface.ipv6_bindings.emplace_back(std::move(b));
                    } else {
                        net_interface.ipv4_bindings.emplace_back(std::move(b));
                    }

                    // http://support.microsoft.com/kb/894564 talks about how binding order is determined.
                    // GetAdaptersAddresses returns adapters in binding order. This way, the domain and primary_interface match.
                    // The old facter behavior didn't make a lot of sense (it would pick the last in binding order, not 1st).
                    // Only accept this as a primary interface if it has a non-link-local address.
                    if (result.primary_interface.empty() && (
                        (it->Address.lpSockaddr->sa_family == AF_INET && !ignored_ipv4_address(addr)) ||
                        (it->Address.lpSockaddr->sa_family == AF_INET6 && !ignored_ipv6_address(addr)))) {
                        result.primary_interface = net_interface.name;
                    }
                }
            }

            stringstream macaddr;
            for (DWORD i = 0u; i < pCurAddr->PhysicalAddressLength; ++i) {
                macaddr << setfill('0') << setw(2) << hex << uppercase <<
                    static_cast<int>(pCurAddr->PhysicalAddress[i]) << ':';
            }
            net_interface.macaddress = macaddr.str();
            if (!net_interface.macaddress.empty()) {
                net_interface.macaddress.pop_back();
            }

            net_interface.mtu = pCurAddr->Mtu;

            result.interfaces.emplace_back(move(net_interface));
        }

        return result;
    }

}}}  // namespace facter::facts::windows

File: lib/resolvers/windows/networking_resolver.rb
# frozen_string_literal: true

require 'ipaddr'

module Facter
  module Resolvers
    class Networking < BaseResolver
      @log = Facter::Log.new(self)
      @semaphore = Mutex.new
      @fact_list ||= {}
      class << self
        private

        def post_resolve(fact_name)
          @fact_list.fetch(fact_name) { read_network_information(fact_name) }
        end

        def read_network_information(fact_name)
          require "#{ROOT_DIR}/lib/resolvers/windows/ffi/networking_ffi"

          size_ptr = FFI::MemoryPointer.new(NetworkingFFI::BUFFER_LENGTH)
          adapter_addresses = FFI::MemoryPointer.new(IpAdapterAddressesLh.size, NetworkingFFI::BUFFER_LENGTH)
          flags = NetworkingFFI::GAA_FLAG_SKIP_ANYCAST |
                  NetworkingFFI::GAA_FLAG_SKIP_MULTICAST | NetworkingFFI::GAA_FLAG_SKIP_DNS_SERVER

          return unless (adapter_addresses = get_adapter_addresses(size_ptr, adapter_addresses, flags))

          iterate_list(adapter_addresses)
          set_interfaces_other_facts if @fact_list[:interfaces]
          @fact_list[:primary_interface] = @fact_list[:primary_interface].to_s
          @fact_list[fact_name]
        end

        def get_adapter_addresses(size_ptr, adapter_addresses, flags)
          error = nil
          3.times do
            error = NetworkingFFI::GetAdaptersAddresses(NetworkingFFI::AF_UNSPEC, flags,
                                                        FFI::Pointer::NULL, adapter_addresses, size_ptr)
            break if error == NetworkingFFI::ERROR_SUCCES

            if error == NetworkingFFI::ERROR_BUFFER_OVERFLOW
              adapter_addresses = FFI::MemoryPointer.new(IpAdapterAddressesLh.size, NetworkingFFI::BUFFER_LENGTH)
            else
              @log.debug 'Unable to retrieve networking facts!'
              return nil
            end
          end
          return nil unless error.zero?

          adapter_addresses
        end

        def adapter_down?(adapter)
          adapter[:OperStatus] != NetworkingFFI::IF_OPER_STATUS_UP ||
            ![NetworkingFFI::IF_TYPE_ETHERNET_CSMACD, NetworkingFFI::IF_TYPE_IEEE80211].include?(adapter[:IfType])
        end

        def retrieve_dhcp_server(adapter)
          if !(adapter[:Flags] & NetworkingFFI::IP_ADAPTER_DHCP_ENABLED).zero? &&
             adapter[:Union][:Struct][:Length] >= IpAdapterAddressesLh.size
            NetworkUtils.address_to_string(adapter[:Dhcpv4Server])
          end
        end

        def iterate_list(adapter_addresses)
          net_interface = {}
          IpAdapterAddressesLh.read_list(adapter_addresses) do |adapter_address|
            if adapter_down?(adapter_address)
              adapter_address = IpAdapterAddressesLh.new(adapter_address[:Next])
              next
            end
            @fact_list[:domain] ||= adapter_address[:DnsSuffix].read_wide_string_without_length
            name = adapter_address[:FriendlyName].read_wide_string_without_length.to_sym
            net_interface[name] = build_interface_info(adapter_address, name)
          end

          @fact_list[:interfaces] = net_interface unless net_interface.empty?
        end

        def build_interface_info(adapter_address, name)
          hash = {}

          hash[:dhcp] = retrieve_dhcp_server(adapter_address)
          hash[:mtu] = adapter_address[:Mtu]

          bindings = find_ip_addresses(adapter_address[:FirstUnicastAddress], name)
          hash[:bindings] = bindings[:ipv4] unless bindings[:ipv4].empty?
          hash[:bindings6] = bindings[:ipv6] unless bindings[:ipv6].empty?
          hash[:mac] = NetworkUtils.find_mac_address(adapter_address)
          hash
        end

        def find_ip_addresses(unicast_addresses, name)
          bindings = {}
          bindings[:ipv6] = []
          bindings[:ipv4] = []

          IpAdapterUnicastAddressLH.read_list(unicast_addresses) do |unicast|
            addr = NetworkUtils.address_to_string(unicast[:Address])
            unless addr
              unicast = IpAdapterUnicastAddressLH.new(unicast[:Next])
              next
            end

            sock_addr = SockAddr.new(unicast[:Address][:lpSockaddr])
            add_ip_data(addr, unicast, sock_addr, bindings)
            find_primary_interface(sock_addr, name, addr)
          end
          bindings
        end

        def add_ip_data(addr, unicast, sock_addr, bindings)
          result = find_bindings(sock_addr, unicast, addr)
          return unless result

          bindings[:ipv6] << result if sock_addr[:sa_family] == NetworkingFFI::AF_INET6
          bindings[:ipv4] << result if sock_addr[:sa_family] == NetworkingFFI::AF_INET
        end

        def find_bindings(sock_addr, unicast, addr)
          return unless [NetworkingFFI::AF_INET, NetworkingFFI::AF_INET6].include?(sock_addr[:sa_family])

          ::Resolvers::Utils::Networking.build_binding(addr, unicast[:OnLinkPrefixLength])
        end

        def find_primary_interface(sock_addr, name, addr)
          if !@fact_list[:primary_interface] &&
             ([NetworkingFFI::AF_INET, NetworkingFFI::AF_INET6].include?(sock_addr[:sa_family]) &&
                 !NetworkUtils.ignored_ip_address(addr))
            @fact_list[:primary_interface] = name
          end
        end

        def set_interfaces_other_facts
          @fact_list[:interfaces].each do |interface_name, value|
            if value[:bindings]
              binding = find_valid_binding(value[:bindings])
              populate_interface(binding, value)
            end
            if value[:bindings6]
              binding = find_valid_binding(value[:bindings6])
              populate_interface(binding, value, true)
            end
            set_networking_other_facts(value, interface_name)
          end
        end

        def find_valid_binding(bindings)
          bindings.each do |binding|
            return binding unless NetworkUtils.ignored_ip_address(binding[:address])
          end
          bindings.empty? ? nil : bindings.first
        end

        def populate_interface(bind, interface, ipv6 = false)
          return if !bind || bind.empty?

          if ipv6
            interface[:ip6] = bind[:address]
            interface[:netmask6] = bind[:netmask]
            interface[:network6] = bind[:network]
            interface[:scope6] =  ::Resolvers::Utils::Networking.get_scope(bind[:address])
          else
            interface[:network] = bind[:network]
            interface[:netmask] = bind[:netmask]
            interface[:ip] = bind[:address]
          end
        end

        def set_networking_other_facts(value, interface_name)
          return unless @fact_list[:primary_interface] == interface_name

          %i[mtu dhcp mac ip ip6 scope6 netmask netmask6 network network6].each do |key|
            @fact_list[key] = value[key]
          end
        end
      end
    end
  end
end
